src/DiffTests.jl

```diff
@@ -25,15 +25,15 @@ names, one should not expect these functions to be "correct" for their original
 num2num_1(x) = sin(x)^2 / cos(x)^2
 num2num_2(x) = 2*x + sqrt(x*x*x)
 num2num_3(x) = 10.31^(x + x) - x
-num2num_4(x) = 1
+num2num_4(x) = one(x)
 num2num_5(x) = 1. / (1. + exp(-x))
 
-const NUMBER_TO_NUMBER_FUNCS = (num2num_1, num2num_2, num2num_3,
-                                num2num_4, num2num_5, identity)
+const NUMBER_TO_NUMBER_FUNCS = [num2num_1, num2num_2, num2num_3,
```

Comment:
For consistency with matrix2matrix functions, which are extended in DiffTestsMKLSparseExt. I can revert num2num to be tuples again.

But out of curiosity - why they were tuples in the first place? What kind of optimisation does it trigger?

Reply:
I assume because tuples reduces allocations.

I don't think it matters much but I'd suggest reverting these changes in the PR because 1) they seem unrelated and 2) it is not clear that they improve anything (on the contrary, they increase allocations).

Reply:
I can revert it for the sets that are not modified in the extension. BTW, how much would it reduce memory overhead (considering that in return Julia would have to construct a tuple type)?

Reply:
:shrug:

Reply:
I'm a fan of keeping changes minimal and avoiding unrelated changes. So I'd suggest reverting these changes, regardless of whether they can/should be changed or not. For consistency, I think it would make sense to just use a tuple in the extension as well - and to open an issue or separate PR if you would like to change these tuples to arrays.

Reply:
It is possible to replace a tuple with the extended one (although I would then have to drop `const` from the declaration); the code would look a little bit more convoluted.
But if you mean that the extension should define its own set of sparse functions to test -- that would require adding support for that set of functions in all AD backend tests suites.

I am also a fan of minimal changes, but there are also consistency and maintainability considerations. For this PR I thought it would be strange to change some function sets into arrays and keep others as tuples, or rewrite tuples from the extension. But if you, as a maintainer, prefer minimal changes, I can minimize them.

Reply:
Oh, I missed that the extension modifies arrays in the package. That does not seem OK to me, I don't think extensions are supposed to do that. For instance, this means that the resulting state depends on the order in which (possibly additional) extensions are loaded etc.

So in case you want to add an extension, I think a different design is needed anyway.

And yes, I very much would prefer more minimal changes.

Reply:
> That does not seem OK to me, I don't think extensions are supposed to do that. For instance, this means that the resulting state depends on the order in which (possibly additional) extensions are loaded etc.

What is the specific scenario that you think would be problematic?
The extensions would not remove functions from the lists, and they would be checking if the function is already in the list.
As for the order -- I don't think the order of functions in the list has any noticeable effects.
One thing to keep in mind is *Revise.jl*, but at the moment *DiffTests.jl* declares function lists as *const*, so *Revise.jl* cannot update current lists on the fly -- there is no regression here.

Also vector-based approach is used (see below) to dynamically populate the list based on Julia version (e.g. because some linear algebra routines are not available in the earlier versions). DiffTests.jl may drop support for Julia 1.0, but in the long run it might be beneficial to have some straightforward mechanism for supporting multiple Julia versions.

And let me refer to the original issue that I am trying to solve -- I've created JuliaDiff/ForwardDiff.jl#589 to address IMO the very relevant practical problem -- improving autodiff efficiency for linear models and multivariate normal distributions with dense or sparse matrices. So the test functions that I am adding here via the extension should be relevant for all autodiff backends. Vanilla *SparseArrays.jl* does not, at the moment, support *ldiv!()* for triangular sparse matrices, hence *MKLSparse.jl* weak dependency.

Reply:
Extensions are just not designed for such a use case - their sole purpose is to extend existing functions.

A problematic scenario here would be another extension that also adds functions to the same array. Then the order of the functions in the array after loading the extensions can be completely arbitrary, and even worse if precompilation and loading of the extensions happens with multiple tasks in parallel two `push!`es might interfere.